_posts/2018-09-06-fullstack-typed-api-with-typescript.md

```diff
@@ -0,0 +1,427 @@
+to implementations:
+
+```ts
+const customers: Customer[] = [];
+let id: number = 1;
+
+const api: ApiMap = {
+    customers: {
+        GET: () => new Promise(resolve => resolve(customers)),
+        POST: (customerName: string) => {
+            let newCustomer = {name: customerName, id: id++};
+            customers.push(newCustomer);
+            return new Promise(resolve => resolve(newCustomer));
+        }
+    },
+    search: {
+        GET: (params: {id?: number, name?: string}) => {
+            let filteredCustomers = customers
+                .filter(customer => !params.id || customer.id == params.id)
+                .filter(customer => !params.name || customer.name == params.name);
+            return new Promise(resolve => resolve(filteredCustomers))
+        }
+    }
+};
+
+// ...
+
+const app = express();
+app.get("/customers", (req, res) => {
+    api.customers.GET().then(msg => {
+        res.send({message: msg});
+    });
+});
+app.post("/customers", (req, res) => {
+    api.customers.POST(req.body.message).then(msg => {
+        res.send({message: msg});
+    });
+});
+app.get("/search", (req, res) => {
+    api.search.GET(req.query).then(msg => {
+        res.send({message: msg});
+    });
+});
+```
+
+If the client tries to call the API with the wrong parameter...
+
+```ts
+api.customers.POST(123);
+```
+
+... the compiler notices this:
+
+```
+error TS2345: Argument of type '123' is not assignable to parameter of type 'string'.
+```
+
+However, we have a problem. The implementations are not completely type checked
+to match the API! In fact, there is a small typo in the client side
+implementation - can you spot it? In addition, there's a lot of duplicated code
+we'd like to avoid. The approach seems to be a good one, but the actual
+implementation needs some work.
+
+Instead of being a type, what if the API was a value? Then we could
+automatically generate code for client fetches and server binds by iterating
+over the API. Turns out, we can, and we also get to maintain type safety.
+
+## Make the object out of you... er, the API
+
+Let's think for a moment. What do we actually want? We want for
+``customers.GET`` to be a function ``() => Promise<Customer[]>`` that is
+located in ``api.customers.GET``. Let's make helper function ``zero<T>(returns:
+T): (T) => Promise<T>`` which, when called with ``zero<Customer[]>([{id: 0,
+name: ''}])``, returns something which has the type ``() =>
+Promise<Customer>``.
+
+```ts
+function zero<R>(returns: R) {
+    return undefined as any as () => Promise<R>;
+}
+```
+...and do the same for ``one()``, which, when called with ``one<string,
+Customer>('', {id: 0, name: ''})``, returns something which has the type ``(p:
+string) => Promise<Customer>``.
+
+```ts
+function one<T, R>() {
+    return undefined as any as (takes: T) => Promise<R>;
+}
+```
+
+Using the new syntax introduced in TypeScript 3.0, we can combine these
```

Comment:
It's unclear (at least to me) which bit of new syntax you're referring to here. Maybe add a reference? Not absolutely necessary, though.